### PR TITLE
Add support for Automatic Migrations

### PR DESCRIPTION
**Pull request by Abizern**

### CODE REVIEW

**Comment by Abizern** [comment]
This has been tested, but I don't want to create a pull request until you create a branch called `tests`.

File: BSManagedDocument.m [hunk 1]
```diff
@@ -136,13 +136,28 @@ - (BOOL)configurePersistentStoreCoordinatorForURL:(NSURL *)storeURL
...
     if (![storeOptions objectForKey:NSReadOnlyPersistentStoreOption])
     {
-        NSMutableDictionary *mutableOptions = [NSMutableDictionary dictionaryWithDictionary:storeOptions];
         [mutableOptions setObject:@NO forKey:NSReadOnlyPersistentStoreOption];
+        optionsModified = YES;
+    }
+
+    if ([self enableAutomaticMigration]) {
+        [mutableOptions setObject:@YES forKey:NSMigratePersistentStoresAutomaticallyOption];
+        [mutableOptions setObject:@YES forKey:NSInferMappingModelAutomaticallyOption];
```
Comment (mikeabdullah) [inline BSManagedDocument.m]: One thing that bothers me is this currently overrides such options if a subclass were to pass them in

Comment (Abizern) [inline BSManagedDocument.m]: In which case override `enableAutomaticMigration` to return `NO` in the subclass and pass in whatever you want. The idea is to create a sensible default.

**Comment by mikeabdullah** [comment]
> I don't want to create a pull request until you create a branch called tests.

Voila, one tests branch.

**Comment by mikeabdullah** [comment]
I'm still struggling to decide if I actually want to have this baked in as a feature. In the meantime, have made some changes on my `automatic-migration` branch which hopefully you're happy to pull back into yours

**Comment by Abizern** [comment]
Nice, thanks.

**Comment by mikeabdullah** [comment]
So I guess my big issue is that this is only really useful during initial development of an app. Once you start shipping to real customers, it seems rather rude to silently upgrade documents upon simply opening them. Instead you want some sort of UI prompt, or to defer the migration until first save, I think.

Come to think of it, I wonder what happens when combined with Versions. Documents opened from browsing through Versions are read-only so migration would be impossible. I wonder if Core Data ignores the request though, or fails with an error.

It seems to me it would be better just to whack a snippet of code into the README that devs can quickly paste into their `BSManagedDocument`-subclass in order to switch on automatic migration. How about you?

**Comment by Abizern** [comment]
I see your point.

You have more experience that I about creating general use frameworks - so I see your point. My current use case is because I'm in development, I I just want to be able to migrate my documents without worrying about them. So currently this works for me.

That said - I think there should be some built in support for migrations. Maybe even a delegate protocol that can be queried when a the model changes.

**Comment by mikeabdullah** [comment]
> Maybe even a delegate protocol that can be queried when a the model changes.

I'm confused. You'd set up some custom object as the document's delegate? That seems freaky. Or something else?

**Comment by Abizern** [comment]
I was thinking out loud. 

I have an idea. If it works I'll push it for comment. 

## 

Abizer

**Pull request closed without merging**